streams/shared/src/main/scala/zio/stream/experimental/ZTransducer.scala

```diff
@@ -124,6 +128,223 @@ object ZTransducer {
+        }
+      }
+    }
+
+  /**
+   * Creates a transducer that folds elements of type `I` into a structure
+   * of type `O` until `max` elements have been folded.
+   *
+   * Like [[foldWeighted]], but with a constant cost function of 1.
+   */
+  def foldUntil[I, O](z: O, max: Long)(f: (O, I) => O): ZTransducer[Any, Nothing, I, O] =
+    foldWeighted[I, O](z)(_ => 1, max)(f)
+
+  /**
+   * Creates a transducer that effectfully folds elements of type `I` into a structure
+   * of type `O` until `max` elements have been folded.
+   *
+   * Like [[foldWeightedM]], but with a constant cost function of 1.
+   */
+  def foldUntilM[R, E, I, O](z: O, max: Long)(f: (O, I) => ZIO[R, E, O]): ZTransducer[R, E, I, O] =
+    foldWeightedM[R, E, I, O](z)(_ => UIO.succeedNow(1), max)(f)
+
+  /**
+   * Creates a transducer that folds elements of type `I` into a structure
+   * of type `O`, until `max` worth of elements (determined by the `costFn`)
+   * have been folded.
+   *
+   * @note Elements that have an individual cost larger than `max` will
+   * cause the stream to hang. See [[foldWeightedDecompose]] for
+   * a variant that can handle these.
+   */
+  def foldWeighted[I, O](z: O)(costFn: I => Long, max: Long)(f: (O, I) => O): ZTransducer[Any, Nothing, I, O] =
+    foldWeightedDecompose[I, O](z)(costFn, max, Chunk.single(_))(f)
+
+  /**
+   * Creates a transducer that folds elements of type `I` into a structure
+   * of type `O`, until `max` worth of elements (determined by the `costFn`)
+   * have been folded.
+   *
+   * The `decompose` function will be used for decomposing elements that
+   * cause an `O` aggregate to cross `max` into smaller elements. For
+   * example:
+   * {{{
+   * Stream(1, 5, 1)
+   *  .aggregate(
+   *    ZTransducer
+   *      .foldWeightedDecompose(List[Int]())((i: Int) => i.toLong, 4,
+   *        (i: Int) => Chunk(i - 1, 1)) { (acc, el) =>
+   *        el :: acc
+   *      }
+   *      .map(_.reverse)
+   *  )
+   *  .runCollect
+   * }}}
+   *
+   * The stream would emit the elements `List(1), List(4), List(1, 1)`.
+   * The [[foldWeightedDecomposeM]] allows the decompose function
+   * to return a `ZIO` value, and consequently it allows the transducer
+   * to fail.
+   */
+  def foldWeightedDecompose[I, O](
+    z: O
+  )(costFn: I => Long, max: Long, decompose: I => Chunk[I])(f: (O, I) => O): ZTransducer[Any, Nothing, I, O] =
+    ZTransducer {
+      case class FoldWeightedState(started: Boolean, result: O, cost: Long)
+
+      val initial = FoldWeightedState(false, z, 0)
+
+      @tailrec def go(in: Chunk[I], os0: Chunk[O], state: FoldWeightedState): (Chunk[O], FoldWeightedState) =
+        in.headOption match {
+          case None => os0 -> state
+          case Some(i) =>
+            val total = state.cost + costFn(i)
+            if (total > max) {
+              val is = decompose(i)
+              if (is.isEmpty)
+                go(in.drop(1), os0 + f(state.result, i), initial)
```

Comment:
We can use `Chunk#foldWhile` here too.

Reply:
Sadly not, because `decompose` will spread the current input and it has to be inserted in place. I could change `go` to take a `List[Chunk[I]]` as an input instead, that would save the `is ++ in.drop(1)`.

Reply:
Ah right. This suggests something like a stack into which you push the current chunk being iterated upon. Sounds a bit overkill for this stage of development though.

Reply:
Forget it, I just found a way to do it single-pass 🤦

Reply:
... though I'd suggest to change `decompose: A => Option[(A, A)]`

Reply:
Is that a must? It's inefficient for cases in which you can decompose to more than two elements.

Reply:
not a must no, check out this version: no `drop`, no `++`, but there is a recursive fold.

Reply:
Looks good! The danger here is that it isn't stacksafe, so decompositions that loop will blow that stack. That's a feature in some sense :-)

Reply:
Yes 😆